File: app/src/main/java/com/itsaky/androidide/activities/AboutActivity.java
/*
 *  This file is part of AndroidIDE.
 *
 *  AndroidIDE is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  AndroidIDE is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.itsaky.androidide.activities;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import androidx.appcompat.widget.TooltipCompat;
import androidx.core.util.Pair;

import com.google.android.gms.oss.licenses.OssLicensesMenuActivity;
import com.itsaky.androidide.R;
import com.itsaky.androidide.app.IDEActivity;
import com.itsaky.androidide.databinding.ActivityAboutBinding;
import com.itsaky.androidide.databinding.LayoutAboutItemsBinding;

import java.util.ArrayList;
import java.util.List;

public class AboutActivity extends IDEActivity {

  public static final String JDK_SOURCE = "https://github.com/itsaky/OpenJDK-Android";
  private ActivityAboutBinding binding;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    binding.items.footerText.setText(getFooter());

    setupTranslations();

    OssLicensesMenuActivity.setActivityTitle(getString(R.string.oss_license_title));

    LayoutAboutItemsBinding items = binding.items;
    TooltipCompat.setTooltipText(items.discuss, getString(R.string.discussions_on_telegram));
    TooltipCompat.setTooltipText(items.github, getString(R.string.title_github));
    TooltipCompat.setTooltipText(items.email, getString(R.string.about_option_email));
    TooltipCompat.setTooltipText(items.website, getString(R.string.about_option_website));
    TooltipCompat.setTooltipText(items.licenses, getString(R.string.oss_license_title));
    items.github.setOnClickListener(v -> getApp().openGitHub());
    items.discuss.setOnClickListener(v -> getApp().openTelegramGroup());
    items.email.setOnClickListener(v -> getApp().emailUs());
    items.website.setOnClickListener(v -> getApp().openWebsite());
    items.licenses.setOnClickListener(
        v -> startActivity(new Intent(this, OssLicensesMenuActivity.class)));
  }

  @Override
  protected View bindLayout() {
    binding = ActivityAboutBinding.inflate(getLayoutInflater());
    return binding.getRoot();
  }

  private void setupTranslations() {
    final var translations = createTranslationsList();
    final var sb = new StringBuilder();

    for (var translation : translations) {
      final var author = translation.first;
      final var language = translation.second;

      sb.append(author);
      sb.append(" - ");
      sb.append(language);
      sb.append("\n");
    }

    binding.items.translations.setText(sb);
  }

  /**
   * Create the list of translations. The list items are pairs. The first item of the pair is the
   * name of the language. The second item is the name of the person who made the translations.
   *
   * <p>If you're adding translations, make sure you add an item to this list. You can choose to
   * display the language in a localized version. For example : Language: हिन्दी - < name of author
   * >
   *
   * <p>Prefer using unicode escapes for language names. For example, हिन्दी can be written as :
   * Language: \u0939\u093f\u0928\u094d\u0926\u0940 - < name of author >
   *
   * @return The translations list.
   */
  private List<Pair<String, String>> createTranslationsList() {
    final var list = new ArrayList<Pair<String, String>>();

    list.add(Pair.create("Bahasa Indonesia", "Fitrah Nuno Syahbani"));
    list.add(Pair.create("\u4e2d\u6587", "Rosemoe and @mikofe")); // 中文
    list.add(Pair.create("Deutsch", "Marvin Stelter"));
    list.add(Pair.create("\u0939\u093f\u0928\u094d\u0926\u0940", "Premjit Chowdhury")); // हिन्दी
    list.add(Pair.create("Russian", "Smooth-E"));
    list.add(Pair.create("French", "Se-Lyan"));
    list.add(Pair.create("Portuguese-Brazilian", "Frederick'XS"));
    return list;
  }

  private String getFooter() {
    final String arch = android.os.Build.SUPPORTED_ABIS[0];
    try {
      final String version = getPackageManager().getPackageInfo(getPackageName(), 0).versionName;
      return getString(R.string.about_footer, version, arch);
    } catch (Throwable th) {
      return getString(R.string.about_footer_alternate, arch);
    }
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    binding = null;
  }
}

File: core/app/src/main/java/com/itsaky/androidide/activities/AboutActivity.kt
/*
 *  This file is part of AndroidIDE.
 *
 *  AndroidIDE is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  AndroidIDE is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.itsaky.androidide.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.SpannableStringBuilder
import android.text.style.ForegroundColorSpan
import android.view.View
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import androidx.core.graphics.Insets
import androidx.core.view.updatePaddingRelative
import com.blankj.utilcode.util.ClipboardUtils
import com.blankj.utilcode.util.SizeUtils
import com.itsaky.androidide.BuildConfig
import com.itsaky.androidide.R
import com.itsaky.androidide.adapters.SimpleIconTitleDescriptionAdapter
import com.itsaky.androidide.app.BaseApplication
import com.itsaky.androidide.app.EdgeToEdgeIDEActivity
import com.itsaky.androidide.app.configuration.IDEBuildConfigProvider
import com.itsaky.androidide.buildinfo.BuildInfo
import com.itsaky.androidide.databinding.ActivityAboutBinding
import com.itsaky.androidide.models.IconTitleDescriptionItem
import com.itsaky.androidide.models.SimpleIconTitleDescriptionItem
import com.itsaky.androidide.utils.BuildInfoUtils
import com.itsaky.androidide.utils.flashSuccess
import com.itsaky.androidide.utils.resolveAttr

class AboutActivity : EdgeToEdgeIDEActivity() {

  private var _binding: ActivityAboutBinding? = null
  private val binding: ActivityAboutBinding
    get() = checkNotNull(_binding) {
      "Activity has been destroyed"
    }

  override fun bindLayout(): View {
    _binding = ActivityAboutBinding.inflate(layoutInflater)
    return _binding!!.root
  }

  companion object {

    private var id = 0
    private val ACTION_WEBSITE = id++
    private val ACTION_EMAIL = id++
    private val ACTION_TG_CHANNEL = id++
    private val ACTION_TG_GROUP = id++
    private val ACTION_CONTRIBUTE = id++
    private val ACTION_CONTRIBUTORS = id++
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    binding.apply {

      setSupportActionBar(toolbar)
      supportActionBar!!.setDisplayHomeAsUpEnabled(true)
      supportActionBar!!.setTitle(R.string.about)
      toolbar.setNavigationOnClickListener { onBackPressedDispatcher.onBackPressed() }

      aboutHeader.apply {
        ideVersion.text = createVersionText()
        ideVersion.isClickable = true
        ideVersion.isFocusable = true
        ideVersion.setBackgroundResource(R.drawable.bg_ripple)
        ideVersion.setOnClickListener {
          ClipboardUtils.copyText(BuildInfoUtils.getBuildInfoHeader())
          flashSuccess(R.string.copied)
        }
      }

      socials.apply {
        sectionTitle.setText(R.string.title_socials)
        sectionItems.adapter = AboutSocialItemsAdapter(createSocialItems(), ::handleActionClick)
      }

      misc.apply {
        sectionTitle.setText(R.string.title_misc)
        sectionItems.adapter = AboutSocialItemsAdapter(createMiscItems(), ::handleActionClick)
      }
    }
  }

  override fun onApplySystemBarInsets(insets: Insets) {
    binding.toolbar.apply {
      setPaddingRelative(
        paddingStart + insets.left,
        paddingTop,
        paddingEnd + insets.right,
        paddingBottom
      )
    }
  }

  private fun handleActionClick(action: SimpleIconTitleDescriptionItem) {
    when (action.id) {
      ACTION_WEBSITE -> app.openWebsite()
      ACTION_EMAIL -> app.emailUs()
      ACTION_TG_GROUP -> app.openTelegramGroup()
      ACTION_TG_CHANNEL -> app.openTelegramChannel()
      ACTION_CONTRIBUTE -> app.openUrl(BaseApplication.CONTRIBUTOR_GUIDE_URL)
      ACTION_CONTRIBUTORS -> startActivity(Intent(this, ContributorsActivity::class.java))
    }
  }

  private fun createSocialItems(): List<IconTitleDescriptionItem> {
    return mutableListOf<IconTitleDescriptionItem>().apply {
      add(
        createSimpleIconTextItem(
          this@AboutActivity,
          ACTION_WEBSITE,
          R.drawable.ic_website,
          R.string.about_option_website,
          BuildInfo.PROJECT_SITE
        )
      )
      add(
        createSimpleIconTextItem(
          this@AboutActivity,
          ACTION_EMAIL,
          R.drawable.ic_email,
          R.string.about_option_email,
          BaseApplication.EMAIL
        )
      )
      add(
        createSimpleIconTextItem(
          this@AboutActivity,
          ACTION_TG_GROUP,
          R.drawable.ic_telegram,
          R.string.discussions_on_telegram,
          BaseApplication.TELEGRAM_GROUP_URL
        )
      )
      add(
        createSimpleIconTextItem(
          this@AboutActivity,
          ACTION_TG_CHANNEL,
          R.drawable.ic_telegram,
          R.string.official_tg_channel,
          BaseApplication.TELEGRAM_CHANNEL_URL
        )
      )
    }
  }

  private fun createMiscItems(): List<IconTitleDescriptionItem> {
    return mutableListOf<IconTitleDescriptionItem>().apply {
      add(
        SimpleIconTitleDescriptionItem.create(
          this@AboutActivity,
          ACTION_CONTRIBUTE,
          R.drawable.ic_code,
          R.string.title_contribute,
          R.string.summary_contribute
        )
      )
      add(
        SimpleIconTitleDescriptionItem.create(
          this@AboutActivity,
          ACTION_CONTRIBUTORS,
          R.drawable.ic_heart_outline,
          R.string.title_contributors,
          R.string.summary_contributors
        )
      )
    }
  }

  private fun createSimpleIconTextItem(
    context: Context,
    id: Int,
    @DrawableRes icon: Int,
    @StringRes title: Int,
    description: CharSequence
  ): SimpleIconTitleDescriptionItem {
    return SimpleIconTitleDescriptionItem(
      id,
      ContextCompat.getDrawable(context, icon),
      ContextCompat.getString(context, title),
      description
    )
  }

  /**
   * Create the version name string that should be displayed to the user.
   *
   * Format of the version name string is :
   *
   * `v[version-name]-[variant] ([build-type]/[[UN]OFFICIAL])`
   */
  @Suppress("KDocUnresolvedReference")
  private fun createVersionText(): CharSequence {
    val builder = SpannableStringBuilder()
    builder.append("v")
    builder.append(BuildInfo.VERSION_NAME_SIMPLE)
    builder.append("-")
    builder.append(IDEBuildConfigProvider.getInstance().cpuAbiName)
    builder.append(" ")

    val colorPositive = ContextCompat.getColor(this, R.color.color_success)
    val colorNegative = ContextCompat.getColor(this, R.color.color_error)

    appendBuildType(builder, colorPositive, colorNegative)

    return builder
  }

  private fun appendBuildType(
    builder: SpannableStringBuilder,
    @ColorInt
    colorPositive: Int,
    @ColorInt
    colorNegative: Int
  ) {
    @Suppress("KotlinConstantConditions")
    var color = if (BuildConfig.BUILD_TYPE != "release") {
      colorNegative
    } else {
      colorPositive
    }

    builder.append("(")
    appendForegroundSpan(builder, BuildConfig.BUILD_TYPE, color)

    val isOfficialBuild = BuildInfoUtils.isOfficialBuild(this)

    color = if (isOfficialBuild) {
      colorPositive
    } else {
      colorNegative
    }

    builder.append("/")
    appendForegroundSpan(
      builder,
      BuildInfoUtils.getBuildType(this).lowercase(),
      color
    )

    builder.append(")")
  }

  private fun appendForegroundSpan(
    builder: SpannableStringBuilder,
    text: CharSequence,
    color: Int
  ) {
    builder.append(
      text,
      ForegroundColorSpan(color),
      SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE
    )
  }

  override fun onDestroy() {
    super.onDestroy()
    _binding = null
  }

  class AboutSocialItemsAdapter(
    items: List<IconTitleDescriptionItem>,
    private val onClickListener: (SimpleIconTitleDescriptionItem) -> Unit
  ) : SimpleIconTitleDescriptionAdapter(items) {

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
      super.onBindViewHolder(holder, position)
      val binding = holder.binding
      val item = getItem(position) as SimpleIconTitleDescriptionItem
      val dp8 = SizeUtils.dp2px(8f)
      binding.icon.updatePaddingRelative(dp8, dp8, dp8, dp8)
      binding.title.setTextAppearance(R.style.TextAppearance_Material3_TitleSmall)

      binding.description.maxLines = 3
      binding.description.setTextAppearance(R.style.TextAppearance_Material3_BodySmall)
      binding.description.setTextColor(binding.description.context.resolveAttr(R.attr.colorPrimary))

      binding.root.isClickable = true
      binding.root.isFocusable = true
      binding.root.setBackgroundResource(R.drawable.bg_ripple)
      binding.root.setOnClickListener {
        onClickListener(item)
      }
    }
  }
}
